### PR TITLE
Add images tab

### PR DESCRIPTION
**Pull request by jyoung-gds**
Trello ticket [link 1](https://trello.com/c/0M9vyh5i/1106-create-separate-images-page), [link 2](https://trello.com/c/AvguABf6/1056-new-image-tab), [link 3](https://trello.com/c/koVf63dk/1118-create-separate-images-page)

## What
This PR creates a separate resource for images, adding a controller and an index endpoint. It then uses this endpoint when creating a tab for adding/removing/editing images on the edit edition page, which is hidden behind a feature flag. 

**NOTE:** None of the UI is functional at the moment, this is just a backbone for the upcoming update to images workflow.

## Why
This is required for the upcoming work on updating the images workflow.

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.

### CODE REVIEW

File: config/routes.rb
Comment (davidgisbey) [inline config/routes.rb:273]: generally when adding endpoints it's good to just add the ones you need as you need them. It might be that you end up needing new/create/edit/update etc., but for now i'd just add 

```
resources :images, only: %i[index]
```

File: app/controllers/admin/images_controller.rb
Comment (davidgisbey) [inline app/controllers/admin/images_controller.rb:23]: i think we need to enforce a similar permission to the edition summary page. This is from the editions controller

```
  def enforce_permissions!
    case action_name
    when "index", "topics", "history"
      enforce_permission!(:see, edition_class || Edition)
    when "show"
      enforce_permission!(:see, @edition)
    when "new"
      enforce_permission!(:create, edition_class || Edition)
    when "create"
      enforce_permission!(:create, @edition)
    when "edit", "update", "revise", "diff", "update_bypass_id"
      enforce_permission!(:update, @edition)
    when "destroy", "confirm_destroy"
      enforce_permission!(:delete, @edition)
    when "export", "confirm_export"
      enforce_permission!(:export, edition_class || Edition)
    else
      raise Whitehall::Authority::Errors::InvalidAction, action_name
    end
  end
```

So i think we just need 

```
enforce_permission!(:see,  Edition)
```

then for the new/edit/create/update actions later it would be 

```
:create, Edition
```

File: app/controllers/admin/images_controller.rb
Comment (davidgisbey) [inline app/controllers/admin/images_controller.rb:25]: I think it would be good to move this into a separate before action which lives at the top of the before_actions.

Something like

```
before_action :find_edition
before_action :redirect_unless_user_can_preview_image_update
before_action :enforce_permissions!
before_action :build_blank_image
layout "design_system"


def redirect_unless_user_can_preview_image_update
  redirect_to edit_admin_edition_path(@edition) unless current_user.can_preview_images_update?
end
```

File: app/controllers/admin/images_controller.rb
Comment (davidgisbey) [inline app/controllers/admin/images_controller.rb:2]: Is this controller always going to be coupled to images that belong to an edition? It seems likely that's the case for now as we use mounts for images on other objects (promotional features etc.)

I'm assuming that will be the case here as the designs are so customised for the editions flow (has it's own tab etc.)

Just double checking as `find_edition` couples this controller to editions.

If that is the case though maybe we should call this `Admin::EditionImagesController` so it's a bit more explicit that it's only going to be used for images as lots of other objects have them too

**Comment by davidgisbey** [comment]
Great work Johnny. I've left a couple of comments but no big blockers 🚀

File: app/views/admin/images/_uploaded_images.html.erb
Comment (dnkrj) [inline app/views/admin/images/_uploaded_images.html.erb:15]: Worth using class="govuk-body" for the paragraphs here and below.

**Comment by jyoung-gds** [comment]
@davidgisbey @dnkrj I've pushed some changes up to address both of your comments so this is ready for a re-review. I took out the `build_blank_image` method in the controller because it wasn't needed, it was something I mistakenly left in whilst trying to figure out how images currently work (it may be needed at some point but not right now).

**Comment by davidgisbey** [comment]
Looks great Johnny. Very happy to approve if you'd like to get this merged in and then iterate on it.

File: test/functional/admin/edition_images_controller_test.rb
Comment (ollietreend) [inline test/functional/admin/edition_images_controller_test.rb:4]: It would be nice to cover the `redirect_unless_user_can_preview_images_update` and `enforce_permissions!` behaviour here.

There is some prior art for testing permissions in other controllers, for example the attachments controller test:

https://github.com/alphagov/whitehall/blob/4e087da57a78a4fb29cb0ae5b57db91e142e38e8/test/functional/admin/attachments_controller_test.rb#L442-L446

But I don't think this test will copy-and-paste nicely. 😕  The attachments controller test is a 'functional style' test that inherits from `ActionController::TestCase`. This type of test [is discouraged nowadays](https://api.rubyonrails.org/classes/ActionController/TestCase.html).

https://github.com/alphagov/whitehall/blob/4e087da57a78a4fb29cb0ae5b57db91e142e38e8/test/functional/admin/attachments_controller_test.rb#L3

Whereas the images controller test inherits from `ActionDispatch::IntegrationTest` because it's an 'integration style' test. This is the recommended approach and default Rails behaviour when creating new controller tests, which is how you've ended up with it here.

https://github.com/alphagov/whitehall/blob/f506f75b0d95d132558980da7dfa9c6f08239971/test/functional/admin/edition_images_controller_test.rb#L3

These different types of tests operate in slightly different contexts, so the way you write tests differs too.

I think something like this should work:

```ruby
edition = create(:draft_case_study)
login_as create(:world_editor)
get admin_edition_images_path(edition.id)
assert_equal 403, status
```

**Review by ollietreend — COMMENTED** [review]
Nice work @jyoung-gds 👍🏻 

I've just added a couple of things – one is small, and the other is (hopefully!) small-ish.

**Comment by jyoung-gds** [comment]
@davidgisbey yes the plan is to iterate on this and implement things like showing images currently on an edition etc. once this is merged. I think this will speed things up because it will allow more asynchronous working earlier on. I've written up some trello cards for the remaining work

@ollietreend thanks for review! I'll make those changes today

**Review by ollietreend — APPROVED** [review]
Great work @jyoung-gds ✨🚀

**Pull request merged**
